xdp/src/net.c

```diff
@@ -0,0 +1,336 @@
+        return -1;
+    }
+
+    return 0;
+}
+
+int exchange_addresses (const int ifindex, const char *server_ip, bool is_server,
+                        uint8_t *src_mac, uint8_t *dest_mac,
+                        uint32_t *src_ip, uint32_t *dest_ip)
+{
+    retrieve_local_mac (ifindex, src_mac);
+    retrieve_local_ip (ifindex, src_ip);
+
+    int sock = socket (AF_INET, SOCK_DGRAM, 0);
+    if (sock < 0)
+    {
+        perror ("socket");
+        return -1;
+    }
+
+    // bind to local address port 1234
+    struct sockaddr_in local_addr;
+    local_addr.sin_family = AF_INET;
+    local_addr.sin_port = htons (1234);
+    local_addr.sin_addr.s_addr = INADDR_ANY;
+    int ret = bind (sock, (struct sockaddr *) &local_addr, sizeof (struct sockaddr_in));
+    if (ret < 0)
+    {
+        perror ("bind");
+        return -1;
+    }
+
+    if (!is_server)
+    {
+        // send my info to the server
+        struct sockaddr_in server_addr;
+        server_addr.sin_family = AF_INET;
+        server_addr.sin_port = htons (1234);
+        server_addr.sin_addr.s_addr = inet_addr (server_ip);
+
+        char buf[INFO_PACKET_SIZE];
+        memcpy (buf, src_mac, ETH_ALEN);
+        memcpy (buf + ETH_ALEN, src_ip, sizeof (uint32_t));
+        int ret = sendto (sock, buf, INFO_PACKET_SIZE, 0, (struct sockaddr *) &server_addr, sizeof (struct sockaddr_in));
+        if (ret < 0)
+        {
+            perror ("sendto");
+            return -1;
+        }
+
+        // wait for the server message containing its info
+        memset (buf, 0, INFO_PACKET_SIZE);
+        ret = recvfrom (sock, buf, INFO_PACKET_SIZE, 0, NULL, NULL);
+        if (ret < 0)
+        {
+            perror ("recvfrom");
+            return -1;
+        }
+
+        memcpy (dest_mac, buf, ETH_ALEN);
+        memcpy (dest_ip, buf + ETH_ALEN, sizeof (uint32_t));
+    }
+    else
+    {
+        // wait for the client message containing its info
+        char buf[INFO_PACKET_SIZE];
+        memset (buf, 0, INFO_PACKET_SIZE);
+        ret = recvfrom (sock, buf, INFO_PACKET_SIZE, 0, NULL, NULL);
+        if (ret < 0)
+        {
+            perror ("recvfrom");
+            return -1;
+        }
+
+        memcpy (dest_mac, buf, ETH_ALEN);
+        memcpy (dest_ip, buf + ETH_ALEN, sizeof (uint32_t));
+
+        // send my info to the client
+        struct sockaddr_in client_addr;
+        client_addr.sin_family = AF_INET;
+        client_addr.sin_port = htons (1234);
+        client_addr.sin_addr.s_addr = *dest_ip;
+
+        memcpy (buf, src_mac, ETH_ALEN);
+        memcpy (buf + ETH_ALEN, src_ip, sizeof (uint32_t));
+        ret = sendto (sock, buf, INFO_PACKET_SIZE, 0, (struct sockaddr *) &client_addr, sizeof (struct sockaddr_in));
+        if (ret < 0)
+        {
+            perror ("sendto");
+            return -1;
+        }
+    }
+
+    close (sock);
+
+    return 0;
+}
+
+int build_base_packet (char *buf, const uint8_t *src_mac, const uint8_t *dest_mac,
```

Comment:
What is the reason to go without UDP? You don't have checksum without it, which might be useful for broken packets - do you compensate with it with the `magic` field in the PP proto?

Reply:
The idea was to reduce as much the non-required information transmitted in the packet. 
I think UDP offers no particular advantage: its checksum is handled only by end-hosts and it is checked automatically inside the net stack; however, since our packets bypass it, it must be checked manually which could add delay at UDP-level.
The `magic` field in pingpong payload cannot provide much safety from corruption, since it's a fixed value; also, no "wrong magic number" handling was implemented.
Honestly, I had not even though about packet corruption, which I guess becomes very important when the experiment is run over longer periods of time. If it is likely to be a problem, I see no problem in adding the UDP header.

Reply:
Good point, let's keep it this way for now. We can add checksum logic in the PP payload if we want to add error checking in the future.